File: app/src/lib/editor/image-paths.ts
import { IMAGE_EXTS_ARRAY } from '$lib/utils/mime';
import {
	buildLocalfileUrl,
	isLocalfileUrl,
	stripLocalfilePrefix,
	LOCALFILE_URL_PREFIX
} from '$lib/editor/image-url';
import { DEFAULT_ATTACHMENT_FOLDER } from '$lib/editor/attachments';

/**
 * The one home for converting image destinations between what a note stores on
 * disk (vault-relative, percent-encoded) and what the editor and the WebView
 * need (`localfile://` URLs). Rust used to own the resolve half, which left the
 * synchronous save path and the transclusion path running a second, subtly
 * different implementation; both halves live here now.
 */

const LEGACY_LOCALFILE_PREFIX = 'localfile://localhost';

/** Characters CommonMark accepts bare inside a `(…)` destination. */
const SAFE_DESTINATION = /[A-Za-z0-9\-_.~/]/;

const PERCENT_ESCAPE = /%([0-9A-Fa-f]{2})/g;

/** Decode the percent-escapes in a destination back to the path it names. */
export function decodeDestination(path: string): string {
	if (!path.includes('%')) return path;
	const decoder = new TextDecoder();
	const encoder = new TextEncoder();
	const bytes: number[] = [];
	let last = 0;
	for (const match of path.matchAll(PERCENT_ESCAPE)) {
		bytes.push(...encoder.encode(path.slice(last, match.index)), parseInt(match[1], 16));
		last = (match.index ?? 0) + 3;
	}
	bytes.push(...encoder.encode(path.slice(last)));
	return decoder.decode(Uint8Array.from(bytes));
}

/**
 * Percent-encode a path for a Markdown destination. A space is invalid there
 * and a `)` would end the destination early, so a note whose file name has
 * either would not round-trip. Already-escaped input is decoded first, so
 * encoding the same path twice changes nothing.
 */
export function encodeDestination(path: string): string {
	const bytes = new TextEncoder().encode(decodeDestination(path));
	let out = '';
	for (const byte of bytes) {
		const char = String.fromCharCode(byte);
		out += SAFE_DESTINATION.test(char)
			? char
			: `%${byte.toString(16).toUpperCase().padStart(2, '0')}`;
	}
	return out;
}

/** Convert wiki image embeds `![[file.png]]` to standard markdown syntax. */
export function resolveWikiEmbeds(md: string, attachmentFolder: string | null): string {
	const folder = attachmentFolder || DEFAULT_ATTACHMENT_FOLDER;
	let result = '';
	let pos = 0;

	while (pos < md.length) {
		const start = md.indexOf('![[', pos);
		if (start === -1) {
			result += md.slice(pos);
			break;
		}
		result += md.slice(pos, start);

		const bracket = md.indexOf(']', start + 3);
		if (bracket === -1 || md[bracket + 1] !== ']') {
			result += '![[';
			pos = start + 3;
			continue;
		}

		const filename = md.slice(start + 3, bracket);
		const dot = filename.lastIndexOf('.');
		const ext = dot === -1 ? '' : filename.slice(dot + 1).toLowerCase();
		if (IMAGE_EXTS_ARRAY.includes(ext)) {
			const relPath = filename.includes('/') ? filename : `${folder}/${filename}`;
			result += `![${filename}](${encodeDestination(relPath)})`;
		} else {
			result += md.slice(start, bracket + 2);
		}
		pos = bracket + 2;
	}

	return result;
}

/** Unescape `!\[alt\](url)` — what the serializer emits once a destination with
 * a bare space made markdown-it parse the image as plain text. */
function unescapeImageMarkdown(md: string): string {
	if (!md.includes('!\\[')) return md;
	let result = '';
	let pos = 0;

	while (pos < md.length) {
		const start = md.indexOf('!\\[', pos);
		if (start === -1) {
			result += md.slice(pos);
			break;
		}
		result += md.slice(pos, start);
		const altStart = start + 3;

		// The close may itself be escaped (`\]`), and `\x` pairs must be skipped
		// so an escaped bracket inside the alt text does not end it early.
		let i = altStart;
		let altEnd = -1;
		let skip = 1;
		while (i < md.length) {
			if (md[i] === '\\' && md[i + 1] === ']') {
				altEnd = i;
				skip = 2;
				break;
			}
			if (md[i] === ']') {
				altEnd = i;
				break;
			}
			i += md[i] === '\\' ? 2 : 1;
		}

		const close = altEnd === -1 ? -1 : md.indexOf(')', altEnd + skip + 1);
		if (altEnd !== -1 && md[altEnd + skip] === '(' && close !== -1) {
			result += `![${md.slice(altStart, altEnd)}](${md.slice(altEnd + skip + 1, close)})`;
			pos = close + 1;
			continue;
		}

		result += '!\\[';
		pos = altStart;
	}

	return result;
}

/**
 * Rewrite a legacy `localfile://localhost` image URL to the scheme form this
 * platform's WebView can actually resolve.
 */
function rewriteLegacyLocalfileUrls(md: string): string {
	if (!md.includes(LEGACY_LOCALFILE_PREFIX)) return md;
	return md.replaceAll(LEGACY_LOCALFILE_PREFIX, LOCALFILE_URL_PREFIX);
}

/** Encode the spaces inside any localfile URL (image or link). */
function encodeSpacesInLocalfileUrls(md: string): string {
	if (!md.includes(`](${LEGACY_LOCALFILE_PREFIX}`) && !md.includes(`](${LOCALFILE_URL_PREFIX}`)) {
		return md;
	}
	return mapDestinations(md, (url) => (isLocalfileUrl(url) ? url.replaceAll(' ', '%20') : null));
}

/** Scan `![alt](url)` and rewrite each image for which `transform` returns a
 * replacement. `![[…]]` embeds are left alone — they are not images yet. */
function mapImageLinks(md: string, transform: (alt: string, url: string) => string | null): string {
	let result = '';
	let pos = 0;

	while (pos < md.length) {
		const start = md.indexOf('![', pos);
		if (start === -1) {
			result += md.slice(pos);
			break;
		}

		if (md[start + 2] === '[') {
			result += md.slice(pos, start + 2);
			pos = start + 2;
			continue;
		}

		const bracket = md.indexOf(']', start + 2);
		const urlStart = bracket + 2;
		const close = bracket === -1 ? -1 : md.indexOf(')', urlStart);
		if (close !== -1 && md[bracket + 1] === '(') {
			const replacement = transform(md.slice(start + 2, bracket), md.slice(urlStart, close));
			if (replacement !== null) {
				result += md.slice(pos, start) + replacement;
				pos = close + 1;
				continue;
			}
		}

		result += md.slice(pos, start + 2);
		pos = start + 2;
	}

	return result;
}

/** Scan the destination of every `](…)` and rewrite it where `transform`
 * returns a replacement. Covers plain links as well as images. */
function mapDestinations(md: string, transform: (url: string) => string | null): string {
	let result = '';
	let pos = 0;

	while (pos < md.length) {
		const at = md.indexOf('](', pos);
		if (at === -1) {
			result += md.slice(pos);
			break;
		}
		const urlStart = at + 2;
		const close = md.indexOf(')', urlStart);
		if (close === -1) {
			result += md.slice(pos);
			break;
		}

		const replacement = transform(md.slice(urlStart, close));
		result +=
			replacement === null ? md.slice(pos, close + 1) : md.slice(pos, urlStart) + replacement + ')';
		pos = close + 1;
	}

	return result;
}

/**
 * Convert relative image destinations in markdown to `localfile://` URLs the
 * WebView can load. Repairs what older builds wrote on the way: a
 * `localfile://localhostC:` URL missing its slash, an escaped `![`, and a URL
 * form this platform cannot resolve.
 */
export function resolveImagePaths(md: string, vaultPath: string | null): string {
	if (!vaultPath) return md;

	const slashed = `${LEGACY_LOCALFILE_PREFIX}/`;
	const repaired =
		md.includes(LEGACY_LOCALFILE_PREFIX) && !md.includes(slashed)
			? md.replaceAll(LEGACY_LOCALFILE_PREFIX, slashed)
			: md;
	const prepared = encodeSpacesInLocalfileUrls(
		rewriteLegacyLocalfileUrls(unescapeImageMarkdown(repaired))
	);

	return mapImageLinks(prepared, (alt, url) => {
		if (url === '' || /^(https?:\/\/|data:|localfile:\/\/)/.test(url)) return null;
		// The destination is already the on-disk spelling (percent-encoded); the
		// URL keeps it, so a `(` in a file name never reaches the URL text.
		const absPath = url.startsWith('/') ? `${vaultPath}${url}` : `${vaultPath}/${url}`;
		return `![${alt}](${buildLocalfileUrl(absPath)})`;
	});
}

/** Convert `localfile://` image URLs back to vault-relative destinations. */
export function unresolveImagePaths(md: string, vaultPath: string | null): string {
	if (!vaultPath) return md;

	const vault = normalizeAbsolute(vaultPath);
	return mapImageLinks(md, (alt, url) => {
		const tail = stripLocalfilePrefix(url);
		if (tail === null) return null;
		const absPath = normalizeAbsolute(decodeDestination(tail));
		const relPath = absPath.startsWith(`${vault}/`)
			? absPath.slice(vault.length + 1)
			: absPath.replace(/^\//, '');
		return `![${alt}](${encodeDestination(relPath)})`;
	});
}

/** One spelling of an absolute path: leading slash present, trailing one gone.
 * The localfile URL and the vault root disagree about the leading slash. */
function normalizeAbsolute(path: string): string {
	const withLead = path.startsWith('/') ? path : `/${path}`;
	return withLead.length > 1 ? withLead.replace(/\/+$/, '') : withLead;
}

File: app/src-tauri/src/text_transform/image_paths.rs
use super::{IMAGE_EXTS, LEGACY_LOCALFILE_PREFIX, LOCALFILE_URL_PREFIX};

pub(super) fn resolve_wiki_embeds(md: &str, folder: &str) -> String {
    let mut result = String::with_capacity(md.len());
    let mut pos = 0;

    while pos < md.len() {
        if let Some(start) = md[pos..].find("![[") {
            let abs_start = pos + start;
            result.push_str(&md[pos..abs_start]);

            if let Some(close) = md[abs_start + 3..].find(']') {
                let bracket_pos = abs_start + 3 + close;
                if bracket_pos + 1 < md.len() && md.as_bytes()[bracket_pos + 1] == b']' {
                    let filename = &md[abs_start + 3..bracket_pos];
                    let end_pos = bracket_pos + 2;
                    let mut transformed = false;

                    if let Some(dot_pos) = filename.rfind('.') {
                        let ext = filename[dot_pos + 1..].to_lowercase();
                        if IMAGE_EXTS.contains(&ext.as_str()) {
                            let rel_path = if filename.contains('/') {
                                filename.to_string()
                            } else {
                                format!("{}/{}", folder, filename)
                            };
                            let safe_path = rel_path.replace(' ', "%20");
                            result.push_str(&format!("![{}]({})", filename, safe_path));
                            transformed = true;
                        }
                    }

                    if !transformed {
                        result.push_str(&md[abs_start..end_pos]);
                    }
                    pos = end_pos;
                } else {
                    result.push_str("![[");
                    pos = abs_start + 3;
                }
            } else {
                result.push_str(&md[abs_start..]);
                break;
            }
        } else {
            result.push_str(&md[pos..]);
            break;
        }
    }

    result
}

/// Unescape markdown image syntax like `!\[alt\](url)` → `![alt](url)`.
/// This happens when tiptap-markdown's serializer escapes `[`/`]` characters
/// after the content failed to parse as an image (e.g. URL had literal
/// spaces). Without this step, the image stays permanently broken because
/// the escape makes markdown-it treat it as plain text.
fn unescape_image_markdown(md: &str) -> String {
    let mut result = String::with_capacity(md.len());
    let bytes = md.as_bytes();
    let mut pos = 0;

    while pos < md.len() {
        let escape_start = match md[pos..].find("!\\[") {
            Some(offset) => pos + offset,
            None => {
                result.push_str(&md[pos..]);
                break;
            }
        };

        result.push_str(&md[pos..escape_start]);
        let alt_start = escape_start + 3;

        // Find the first `]` or `\]` that closes the alt text, skipping
        // over nested escape pairs like `\[`.
        let mut i = alt_start;
        let mut alt_end = None;
        let mut skip_len = 0;
        while i < md.len() {
            if bytes[i] == b'\\' && i + 1 < md.len() && bytes[i + 1] == b']' {
                alt_end = Some(i);
                skip_len = 2;
                break;
            }
            if bytes[i] == b']' {
                alt_end = Some(i);
                skip_len = 1;
                break;
            }
            if bytes[i] == b'\\' && i + 1 < md.len() {
                i += 2;
                continue;
            }
            i += 1;
        }

        let (alt_end, after_close) = match alt_end {
            Some(e) => (e, e + skip_len),
            None => {
                result.push_str("!\\[");
                pos = alt_start;
                continue;
            }
        };

        if after_close < md.len() && bytes[after_close] == b'(' {
            if let Some(paren) = md[after_close + 1..].find(')') {
                let url_start = after_close + 1;
                let url_end = url_start + paren;
                let alt = &md[alt_start..alt_end];
                let url = &md[url_start..url_end];
                result.push_str(&format!("![{}]({})", alt, url));
                pos = url_end + 1;
                continue;
            }
        }

        result.push_str("!\\[");
        pos = alt_start;
    }

    result
}

fn encode_spaces_in_localfile_urls(md: &str) -> String {
    let mut result = String::with_capacity(md.len());
    let mut pos = 0;
    while pos < md.len() {
        let next = md[pos..]
            .find("](localfile://")
            .map(|s| (pos + s, "](localfile://".len()))
            .or_else(|| {
                md[pos..]
                    .find("](http://localfile.localhost")
                    .map(|s| (pos + s, "](http://localfile.localhost".len()))
            });
        if let Some((abs_start, _)) = next {
            let url_start = abs_start + 2;
            if let Some(close) = md[url_start..].find(')') {
                let url = &md[url_start..url_start + close];
                result.push_str(&md[pos..url_start]);
                result.push_str(&url.replace(' ', "%20"));
                result.push(')');
                pos = url_start + close + 1;
                continue;
            }
        }
        result.push_str(&md[pos..]);
        break;
    }
    result
}

/// On Windows/Android, rewrite `localfile://localhost/…` URLs inside image
/// markdown to `http://localfile.localhost/…` so WebView2 can resolve them.
#[cfg(any(target_os = "windows", target_os = "android"))]
pub(super) fn rewrite_legacy_localfile_urls(md: &str) -> String {
    if !md.contains("localfile://localhost") {
        return md.to_string();
    }
    md.replace("localfile://localhost", LOCALFILE_URL_PREFIX)
}

#[cfg(not(any(target_os = "windows", target_os = "android")))]
pub(super) fn rewrite_legacy_localfile_urls(md: &str) -> String {
    md.to_string()
}

pub(super) fn resolve_image_paths(md: &str, vault_path: &str) -> String {
    // Fix previously-saved malformed localfile:// URLs where the slash between
    // "localhost" and the drive letter was missing (Windows bug).
    // e.g. localfile://localhostC: → localfile://localhost/C:
    let md = if md.contains("localfile://localhost") && !md.contains("localfile://localhost/") {
        md.replace("localfile://localhost", "localfile://localhost/")
    } else {
        md.to_string()
    };
    let md = unescape_image_markdown(&md);
    let md = rewrite_legacy_localfile_urls(&md);
    let md = encode_spaces_in_localfile_urls(&md);
    let md = md.as_str();

    let mut result = String::with_capacity(md.len());
    let mut pos = 0;

    while pos < md.len() {
        if let Some(start) = md[pos..].find("![") {
            let abs_start = pos + start;

            if abs_start + 2 < md.len() && md.as_bytes()[abs_start + 2] == b'[' {
                result.push_str(&md[pos..abs_start + 2]);
                pos = abs_start + 2;
                continue;
            }

            if let Some(close_bracket) = md[abs_start + 2..].find(']') {
                let alt_end = abs_start + 2 + close_bracket;
                let alt = &md[abs_start + 2..alt_end];

                if alt_end + 1 < md.len() && md.as_bytes()[alt_end + 1] == b'(' {
                    let url_start = alt_end + 2;
                    if let Some(close_paren) = md[url_start..].find(')') {
                        if close_paren > 0 {
                            let url = &md[url_start..url_start + close_paren];
                            if !url.starts_with("http://")
                                && !url.starts_with("https://")
                                && !url.starts_with("data:")
                                && !url.starts_with("localfile://")
                            {
                                result.push_str(&md[pos..abs_start]);
                                let sep = if vault_path.starts_with('/') { "" } else { "/" };
                                let full = format!("{}/{}", vault_path, url).replace(' ', "%20");
                                result.push_str(&format!(
                                    "![{}]({}{}{})",
                                    alt, LOCALFILE_URL_PREFIX, sep, full
                                ));
                                pos = url_start + close_paren + 1;
                                continue;
                            }
                        }
                    }
                }
            }

            result.push_str(&md[pos..abs_start + 2]);
            pos = abs_start + 2;
        } else {
            result.push_str(&md[pos..]);
            break;
        }
    }

    result
}

pub(super) fn unresolve_image_paths(md: &str, vault_path: &str) -> String {
    let mut result = String::with_capacity(md.len());
    let mut pos = 0;

    while pos < md.len() {
        if let Some(start) = md[pos..].find("![") {
            let abs_start = pos + start;

            if abs_start + 2 < md.len() && md.as_bytes()[abs_start + 2] == b'[' {
                result.push_str(&md[pos..abs_start + 2]);
                pos = abs_start + 2;
                continue;
            }

            if let Some(close_bracket) = md[abs_start + 2..].find(']') {
                let alt_end = abs_start + 2 + close_bracket;
                let alt = &md[abs_start + 2..alt_end];

                if alt_end + 1 < md.len() && md.as_bytes()[alt_end + 1] == b'(' {
                    let url_start = alt_end + 2;
                    if let Some(close_paren) = md[url_start..].find(')') {
                        let url = &md[url_start..url_start + close_paren];
                        let stripped = url
                            .strip_prefix("http://localfile.localhost")
                            .or_else(|| url.strip_prefix(LEGACY_LOCALFILE_PREFIX));
                        if let Some(abs_path) = stripped {
                            // On Windows abs_path may be "/C:/Users/..." while
                            // vault_path is "C:/Users/...". Strip the leading /.
                            let norm_abs = abs_path.strip_prefix('/').unwrap_or(abs_path);
                            // Spaces may be %20-encoded in the URL
                            let decoded_abs = norm_abs.replace("%20", " ");
                            let vault_prefix = format!("{}/", vault_path);
                            let rel_path = if decoded_abs.starts_with(&vault_prefix) {
                                decoded_abs[vault_prefix.len()..].to_string()
                            } else if norm_abs.starts_with(&vault_prefix.replace(' ', "%20")) {
                                let enc_prefix = vault_prefix.replace(' ', "%20");
                                norm_abs[enc_prefix.len()..].to_string()
                            } else {
                                decoded_abs
                            };
                            result.push_str(&md[pos..abs_start]);
                            result.push_str(&format!("![{}]({})", alt, rel_path));
                            pos = url_start + close_paren + 1;
                            continue;
                        }
                    }
                }
            }

            result.push_str(&md[pos..abs_start + 2]);
            pos = abs_start + 2;
        } else {
            result.push_str(&md[pos..]);
            break;
        }
    }

    result
}
